components/logins/src/import/fennec_logins.rs

```diff
@@ -0,0 +1,179 @@
+                l.passwordField,
+                IFNULL(l.timesUsed, 0),
+                l.encryptedUsername,
+                l.encryptedPassword,
+                l.guid,
+                sanitize_timestamp(l.timeCreated),
+                l.timeLastUsed,
+                sanitize_timestamp(l.timePasswordChanged),
+                NULL,
+                0, -- is_deleted
+                {new} -- sync_status
+            FROM fennec.logins l
+            WHERE -- Checks copied from `Login::check_valid`.
+            l.hostname != '' AND
+            l.encryptedPassword != '' AND
+            (
+                (l.httpRealm IS NULL AND l.formSubmitURL IS NOT NULL) OR
+                (l.httpRealm IS NOT NULL AND l.formSubmitURL IS NULL)
+            )
+            ",
+        new = SyncStatus::New as u8
+    );
+}
+
+pub fn attached_database<'a>(
+    conn: &'a Connection,
+    path: &Url,
+    db_alias: &'static str,
+) -> Result<ExecuteOnDrop<'a>> {
+    conn.execute_named(
+        "ATTACH DATABASE :path AS :db_alias",
+        named_params! {
+            ":path": path.as_str(),
+            ":db_alias": db_alias,
+        },
+    )?;
+    Ok(ExecuteOnDrop {
+        conn,
+        sql: format!("DETACH DATABASE {};", db_alias),
+    })
+}
+
+/// We use/abuse the mirror to perform our import, but need to clean it up
+/// afterwards. This is an RAII helper to do so.
+///
+/// Ideally, you should call `execute_now` rather than letting this drop
+/// automatically, as we can't report errors beyond logging when running
+/// Drop.
+pub struct ExecuteOnDrop<'a> {
+    conn: &'a Connection,
+    sql: String,
+}
+
+impl<'a> ExecuteOnDrop<'a> {
+    pub fn new(conn: &'a Connection, sql: String) -> Self {
+        Self { conn, sql }
+    }
+
+    pub fn execute_now(self) -> Result<()> {
+        self.conn.execute_batch(&self.sql)?;
+        // Don't run our `drop` function.
+        std::mem::forget(self);
+        Ok(())
+    }
+}
+
+impl Drop for ExecuteOnDrop<'_> {
+    fn drop(&mut self) {
+        if let Err(e) = self.conn.execute_batch(&self.sql) {
+            log::error!("Failed to clean up after import! {}", e);
+            log::debug!("  Failed query: {}", &self.sql);
+        }
+    }
+}
+
+/// If `p` is a file URL, return it, otherwise try and make it one.
+///
+/// Errors if `p` is a relative non-url path, or if it's a URL path
+/// that's isn't a `file:` URL.
+pub fn ensure_url_path(p: impl AsRef<Path>) -> Result<Url> {
+    if let Some(u) = p.as_ref().to_str().and_then(|s| Url::parse(s).ok()) {
+        if u.scheme() == "file" {
+            Ok(u)
+        } else {
+            Err(ErrorKind::IllegalDatabasePath(p.as_ref().to_owned()).into())
+        }
+    } else {
+        let p = p.as_ref();
+        let u = Url::from_file_path(p).map_err(|_| ErrorKind::IllegalDatabasePath(p.to_owned()))?;
+        Ok(u)
+    }
+}
+
+#[inline(never)]
+// Adapted from places::import::common::sql_fns::sanitize_timestamp but works with `i64`.
```

Comment:
Do you think it's worth factoring this out into a shared helper? Also, I wonder...we don't have an "earliest valid timestamp" for logins, like you found the other day, but I wonder if we should share that, then cut our losses and use the current timestamp if we have something earlier. Dates from 1993 or 1970 aren't likely to be helpful for folks.